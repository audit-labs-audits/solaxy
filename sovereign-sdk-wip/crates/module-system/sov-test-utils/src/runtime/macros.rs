/// Generates a bare runtime without implementing the `HasCapabilities` trait.
#[macro_export]
macro_rules! generate_runtime_without_capabilities {
    (
        name: $id:ident,
        modules: [$($module_name:ident : $module_ty:path),* $(,)?],
        operating_mode: $operating_mode:path,
        minimal_genesis_config_type: $minimal_genesis_config_ty:path,
        runtime_trait_impl_bounds: [$($runtime_trait_impl_bounds:tt)*],
        kernel_type: $kernel_type:ty,
        auth_type: $auth:ty,
        auth_call_wrapper: $auth_wrapper:expr
        // optional final comma
        $(,)?
    ) => {
        /// Generated test runtime implementation using the testing framework.
        #[derive(
            Default,
            Clone,
            ::sov_modules_api::Genesis,
            ::sov_modules_api::Hooks,
            ::sov_modules_api::DispatchCall,
            ::sov_modules_api::Event,
            ::sov_modules_api::MessageCodec,
            ::sov_modules_api::macros::CliWallet,
            ::sov_modules_api::macros::RuntimeRestApi,
        )]
        pub struct $id<S: ::sov_modules_api::Spec>  where
        $($runtime_trait_impl_bounds)*
        {
            /// The sequencer registry module.
            pub sequencer_registry: $crate::runtime::SequencerRegistry<S>,
            /// The bank module.
            pub bank: $crate::runtime::Bank<S>,
            /// The accounts module
            pub accounts: $crate::runtime::Accounts<S>,
            /// The uniqueness module
            pub uniqueness: $crate::runtime::Uniqueness<S>,
            /// The attester incentives module.
            pub attester_incentives: $crate::runtime::AttesterIncentives<S>,
            /// The chain state module.
            pub chain_state: $crate::runtime::ChainState<S>,
            /// The blob storage module.
            pub blob_storage: $crate::runtime::BlobStorage<S>,
            /// The prover incentives module.
            pub prover_incentives: $crate::runtime::ProverIncentives<S>,
            $(
                /// An external module [`$module_ty`] of the generated runtime.
                pub $module_name: $module_ty
            ),*
        }

        impl<S: ::sov_modules_api::Spec> $crate::runtime::traits::MinimalGenesis<S> for $id<S>
            where
            $($runtime_trait_impl_bounds)*
        {
            fn sequencer_registry_config(config: &GenesisConfig<S>) -> &<$crate::runtime::SequencerRegistry<S> as ::sov_modules_api::Genesis>::Config {
                &config.sequencer_registry
            }
        }

        impl<S: ::sov_modules_api::Spec> GenesisConfig<S> where
            $($runtime_trait_impl_bounds)*
        {
            #[allow(unused)]
            /// Creates a new [`GenesisConfig`] from a minimal genesis config [`::sov_modules_api::Genesis::Config`].
            pub fn from_minimal_config(minimal_config: $minimal_genesis_config_ty,
                $($module_name: <$module_ty as ::sov_modules_api::Genesis>::Config),*
            ) -> Self {
                Self {
                    sequencer_registry: minimal_config.sequencer_registry,
                    bank: minimal_config.bank,
                    accounts: minimal_config.accounts,
                    uniqueness: minimal_config.uniqueness,
                    chain_state: minimal_config.chain_state,
                    blob_storage: minimal_config.blob_storage,
                    prover_incentives: minimal_config.prover_incentives,
                    attester_incentives: minimal_config.attester_incentives,
                    $(
                        $module_name,
                    )*
                }
            }
        }

        impl<S: ::sov_modules_api::Spec> GenesisConfig<S>
        where
            $($runtime_trait_impl_bounds)*
        {
            #[allow(unused)]
            /// Creates a [`$crate::runtime::GenesisParams`] from a [`GenesisConfig`].
            pub fn into_genesis_params(self) -> $crate::runtime::GenesisParams<Self> {
                $crate::runtime::GenesisParams {
                    runtime: self,
                }
            }
        }

        impl<S> $crate::runtime::Runtime<S> for $id<S>
        where
            S: ::sov_modules_api::Spec,
            ::sov_modules_api::transaction::Transaction::<Self, S>: $crate::sov_universal_wallet::schema::SchemaGenerator,
            <Self as ::sov_modules_api::DispatchCall>::Decodable: $crate::sov_universal_wallet::schema::SchemaGenerator,
            $($runtime_trait_impl_bounds)*
        {
            const CHAIN_HASH: [u8; 32] = [11; 32];

            type GenesisConfig = <Self as ::sov_modules_api::Genesis>::Config;
            type GenesisInput = ();
            type Auth = $auth;

            fn endpoints(api_state: sov_modules_api::rest::ApiState<S>) -> ::sov_modules_api::NodeEndpoints {
                use $crate::sov_rollup_apis::dedup::{DeDupEndpoint, NonceDeDupEndpoint};
                use $crate::sov_rollup_apis::schema::{SchemaEndpoint, StandardSchemaEndpoint};
                use $crate::sov_universal_wallet::schema::{ChainData, Schema};
                use ::sov_modules_api::macros::config_value;
                use ::sov_modules_api::transaction::{Transaction, UnsignedTransaction};
                use ::sov_modules_api::rest::HasRestApi;

                let axum_router = Self::default().rest_api(api_state.clone());
                // Provide an endpoint to return dedup information associated with addresses.
                // Since our runtime is using the uniqueness module we can use the provided `NonceDeDupEndpoint` implementation.
                let dedup_endpoint = NonceDeDupEndpoint::new(api_state.clone());
                let axum_router = axum_router.merge(dedup_endpoint.axum_router());

                let schema = Schema::of_rollup_types_with_chain_data::<
                Transaction<Self, S>,
                UnsignedTransaction<Self, S>,
                <Self as ::sov_modules_api::DispatchCall>::Decodable,
                S::Address,
                >(ChainData {
                    chain_id: config_value!("CHAIN_ID"),
                    chain_name: config_value!("CHAIN_NAME").to_string(),
                })
                .unwrap();

                let schema_endpoint = StandardSchemaEndpoint::new(
                    &schema
                )
                .expect("Failed to initialize StandardSchemaEndpoint");
                let axum_router = axum_router.merge(schema_endpoint.axum_router());

                ::sov_modules_api::NodeEndpoints {
                    axum_router,
                    jsonrpsee_module: ::sov_modules_api::prelude::jsonrpsee::RpcModule::new(()),
                    background_handles: Vec::new(),
                }
            }

            fn genesis_config(_input: &Self::GenesisInput) -> ::sov_modules_api::prelude::anyhow::Result<Self::GenesisConfig> {
                unimplemented!()
            }

            fn operating_mode(genesis: &Self::GenesisConfig) -> ::sov_modules_api::runtime::OperatingMode {
                genesis.chain_state.operating_mode
            }

            fn wrap_call(auth_data: <Self::Auth as sov_modules_api::capabilities::TransactionAuthenticator<S>>::Decodable) -> Self::Decodable {
                $auth_wrapper(auth_data)
            }

            fn allow_unregistered_tx(call: &Self::Decodable) -> bool {
                matches!(
                    call,
                    Self::Decodable::SequencerRegistry($crate::runtime::sov_sequencer_registry::CallMessage::Register {..})
                )
            }
        }


        impl<S> sov_modules_api::capabilities::HasKernel<S> for $id<S> where
            S: ::sov_modules_api::Spec,
            $($runtime_trait_impl_bounds)*
        {

            type Kernel<'a> = $kernel_type;

            fn inner(&mut self) -> sov_modules_api::capabilities::Guard<Self::Kernel<'_>> {
                sov_modules_api::capabilities::Guard::new(Self::Kernel {
                    chain_state: &mut self.chain_state,
                    blob_storage: &mut self.blob_storage,
                })
            }

            fn kernel_with_slot_mapping(&self) -> std::sync::Arc<dyn ::sov_modules_api::capabilities::KernelWithSlotMapping<S>> {
                ::std::sync::Arc::new(self.chain_state.clone())
            }
        }
    }
}

/// Base for generating runtimes.
/// Excludes the TransactionAuthenticator trait to allow custom runtimes like EVM to provide their own
/// implementation.
#[macro_export]
macro_rules! generate_runtime {
    (
        name: $id:ident,
        modules: [$($module_name:ident : $module_ty:path),* $(,)?],
        operating_mode: $operating_mode:path,
        minimal_genesis_config_type: $minimal_genesis_config_ty:path,
        gas_enforcer: $payer_name:ident : $gas_enforcer_ty:ty,
        runtime_trait_impl_bounds: [$($runtime_trait_impl_bounds:tt)*],
        kernel_type: $kernel_type:ty,
        auth_type: $auth:ty,
        auth_call_wrapper: $auth_wrapper:expr
        // optional final comma
        $(,)?
    ) => {
        $crate::generate_runtime_without_capabilities! {
            name: $id,
            modules: [$($module_name : $module_ty),*],
            operating_mode: $operating_mode,
            minimal_genesis_config_type: $minimal_genesis_config_ty,
            runtime_trait_impl_bounds: [$($runtime_trait_impl_bounds)*],
            kernel_type: $kernel_type,
            auth_type: $auth,
            auth_call_wrapper: $auth_wrapper
        }

        impl<S> ::sov_modules_api::capabilities::HasCapabilities<S> for $id<S>
        where
            S: ::sov_modules_api::Spec,
            $($runtime_trait_impl_bounds)*
        {
            type Capabilities<'a> = $crate::runtime::StandardProvenRollupCapabilities<'a, S, &'a mut $gas_enforcer_ty>;

            fn capabilities(&mut self) -> ::sov_modules_api::capabilities::Guard<Self::Capabilities<'_>> {
                ::sov_modules_api::capabilities::Guard::new(
                    $crate::runtime::StandardProvenRollupCapabilities {
                        bank: &mut self.bank,
                        gas_payer: &mut self. $payer_name,
                        sequencer_registry: &mut self.sequencer_registry,
                        accounts: &mut self.accounts,
                        uniqueness: &mut self.uniqueness,
                        prover_incentives: &mut self.prover_incentives,
                        attester_incentives: &mut self.attester_incentives,
                    }
                )
            }

        }
    };
    (
        name: $id:ident,
        modules: [$($module_name:ident : $module_ty:path),* $(,)?],
        operating_mode: $operating_mode:path,
        minimal_genesis_config_type: $minimal_genesis_config_ty:path,
        runtime_trait_impl_bounds: [$($runtime_trait_impl_bounds:tt)*],
        kernel_type: $kernel_type:ty,
        auth_type: $auth:ty,
        auth_call_wrapper: $auth_wrapper:expr
        // optional final comma
        $(,)?
    ) => {
        $crate::generate_runtime_without_capabilities! {
            name: $id,
            modules: [$($module_name : $module_ty),*],
            operating_mode: $operating_mode,
            minimal_genesis_config_type: $minimal_genesis_config_ty,
            runtime_trait_impl_bounds: [$($runtime_trait_impl_bounds)*],
            kernel_type: $kernel_type,
            auth_type: $auth,
            auth_call_wrapper: $auth_wrapper,
        }

        impl<S> ::sov_modules_api::capabilities::HasCapabilities<S> for $id<S>
        where
            S: ::sov_modules_api::Spec,
            $($runtime_trait_impl_bounds)*
        {
            type Capabilities<'a> = $crate::runtime::StandardProvenRollupCapabilities<'a, S>;

            fn capabilities(&mut self) -> ::sov_modules_api::capabilities::Guard<Self::Capabilities<'_>> {
                ::sov_modules_api::capabilities::Guard::new(
                    $crate::runtime::StandardProvenRollupCapabilities {
                        bank: &mut self.bank,
                        gas_payer: (),
                        sequencer_registry: &mut self.sequencer_registry,
                        accounts: &mut self.accounts,
                        uniqueness: &mut self.uniqueness,
                        prover_incentives: &mut self.prover_incentives,
                        attester_incentives: &mut self.attester_incentives,
                    }
                )
            }

        }
    }
}

/// Generates a optimistic runtime containing the [`Bank`](sov_bank::Bank), [`AttesterIncentives`](sov_attester_incentives::AttesterIncentives),
/// and [`SequencerRegistry`](sov_sequencer_registry::SequencerRegistry) modules in addition to any provided as arguments. The runtime implements a basic kernel.
#[macro_export]
macro_rules! generate_optimistic_runtime {
    ($id:ident <= $($module_name:ident : $module_ty:path),*) => {
        $crate::generate_optimistic_runtime_with_kernel! {
            $id <= kernel_type: $crate::runtime::BasicKernel<'a, S>, $($module_name : $module_ty),*
        }
    };
}

/// Generates a optimistic runtime containing the [`Bank`](sov_bank::Bank), [`AttesterIncentives`](sov_attester_incentives::AttesterIncentives),
/// and [`SequencerRegistry`](sov_sequencer_registry::SequencerRegistry) modules in addition to any provided as arguments. The runtime implements a custom kernel.
#[macro_export]
macro_rules! generate_optimistic_runtime_with_kernel {
    ($id:ident <= kernel_type: $kernel_ty:ty, $($module_name:ident : $module_ty:path),*) => {
        $crate::generate_runtime! {
            name: $id,
            modules: [$($module_name : $module_ty),*],
            operating_mode: sov_modules_api::runtime::OperatingMode::Optimistic,
            minimal_genesis_config_type: $crate::runtime::genesis::optimistic::config::MinimalOptimisticGenesisConfig<S>,
            runtime_trait_impl_bounds: [],
            kernel_type: $kernel_ty,
            auth_type: sov_modules_api::capabilities::RollupAuthenticator<S, Self>,
            auth_call_wrapper: |auth_data| auth_data,
        }
    };
}

/// Generates a zk runtime containing the [`Bank`](sov_bank::Bank), [`ProverIncentives`](sov_prover_incentives::ProverIncentives),
/// and [`SequencerRegistry`](sov_sequencer_registry::SequencerRegistry) modules in addition to any provided as arguments. The runtime implements a basic kernel.
#[macro_export]
macro_rules! generate_zk_runtime {
    ($id:ident <= $($module_name:ident : $module_ty:path),*) => {
        $crate::generate_zk_runtime_with_kernel! {
            kernel_type: $crate::runtime::BasicKernel<'a, S>,
            $id <= $($module_name : $module_ty),*
        }
    };
}

/// Generates a zk runtime containing the [`Bank`](sov_bank::Bank), [`ProverIncentives`](sov_prover_incentives::ProverIncentives),
/// and [`SequencerRegistry`](sov_sequencer_registry::SequencerRegistry) modules in addition to any provided as arguments. The runtime implements a custom kernel.
#[macro_export]
macro_rules! generate_zk_runtime_with_kernel {
    (kernel_type: $kernel_ty:ty, $id:ident <= $($module_name:ident : $module_ty:path),*) => {
        $crate::generate_runtime! {
            name: $id,
            modules: [$($module_name : $module_ty),*],
            operating_mode: sov_modules_api::runtime::OperatingMode::Zk,
            minimal_genesis_config_type: $crate::runtime::genesis::zk::MinimalZkGenesisConfig<S>,
            runtime_trait_impl_bounds: [],
            kernel_type: $kernel_ty,
            auth_type: sov_modules_api::capabilities::RollupAuthenticator<S, Self>,
            auth_call_wrapper: |auth_data| auth_data,
        }
    };
}

/// Assert that a pattern matches the expected value.
/// This should be replaced by `std` version when it is stabilized: `<https://github.com/rust-lang/rust/issues/82775>`
#[macro_export]
macro_rules! assert_matches {
    ($value:expr, $pattern:pat) => {
        assert_matches!($value, $pattern, "")
    };
    ($value:expr, $pattern:pat if $guard:expr) => {
        assert_matches!($value, $pattern if $guard, "")
    };
    ($value:expr, $pattern:pat, $message:expr) => {{
        match $value {
            $pattern => (),
            ref _v => panic!(
                "{}Assertion failed:\nExpected: {}\nReceived: {:?}",
                if $message.is_empty() {
                    String::new()
                } else {
                    format!("{}\n", $message)
                },
                stringify!($pattern),
                _v
            ),
        }
    }};
    ($value:expr, $pattern:pat if $guard:expr, $message:expr) => {{
        match $value {
            v @ $pattern => {
                if !($guard) {
                    panic!(
                        "{}Assertion failed:\nExpected: {} if {}\nReceived: {:?}",
                        if $message.is_empty() {
                            String::new()
                        } else {
                            format!("{}\n", $message)
                        },
                        stringify!($pattern),
                        stringify!($guard),
                        v
                    )
                }
            }
            ref _v => panic!(
                "{}Assertion failed:\nExpected: {} if {}\nReceived: {:?}",
                if $message.is_empty() {
                    String::new()
                } else {
                    format!("{}\n", $message)
                },
                stringify!($pattern),
                stringify!($guard),
                _v
            ),
        }
    }};
}
