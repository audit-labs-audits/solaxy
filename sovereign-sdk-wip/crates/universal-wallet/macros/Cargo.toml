[package]
name = "sov-universal-wallet-macros"
version = "0.1.0"
edition = "2021"
license = "MIT OR Apache-2.0"

[lib]
proc-macro = true

[dependencies]
syn = { workspace = true }
proc-macro2 = { workspace = true }
sov-universal-wallet-macro-helpers = { workspace = true }
