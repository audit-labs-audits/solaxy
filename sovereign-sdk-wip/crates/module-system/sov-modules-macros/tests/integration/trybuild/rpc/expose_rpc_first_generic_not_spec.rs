use jsonrpsee::core::RpcResult;
use sov_modules_api::macros::{expose_rpc, rpc_gen};
use sov_modules_api::prelude::UnwrapInfallible;
use sov_modules_api::{
    ApiStateAccessor, Context, DaSpec, DispatchCall, Error, Genesis, MessageCodec, Module,
    ModuleId, ModuleInfo, Spec, StateValue, TxState,
};

pub trait TestSpec:
    Default + Send + Sync + std::fmt::Debug + Clone + PartialEq + Eq + 'static
{
    type Data: Data;
}

pub trait Data:
    Clone
    + Eq
    + PartialEq
    + std::fmt::Debug
    + serde::Serialize
    + serde::de::DeserializeOwned
    + borsh::BorshSerialize
    + borsh::BorshDeserialize
    + schemars::JsonSchema
    + Send
    + Sync
    + 'static
{
}

impl Data for u32 {}

pub mod my_module {
    use super::*;

    #[derive(ModuleInfo, Clone)]
    pub struct QueryModule<S: Spec, D: Data> {
        #[id]
        pub id: ModuleId,

        #[state]
        pub data: StateValue<D>,

        #[phantom]
        phantom: std::marker::PhantomData<S>,
    }

    impl<S: Spec, D> Module for QueryModule<S, D>
    where
        D: Data,
    {
        type Spec = S;
        type Config = D;
        type CallMessage = D;
        type Event = ();

        fn genesis(
            &mut self,
            _genesis_rollup_header: &<S::Da as DaSpec>::BlockHeader,

            config: &Self::Config,
            state: &mut impl sov_modules_api::GenesisState<S>,
        ) -> Result<(), Error> {
            self.data.set(config, state).unwrap();
            Ok(())
        }

        fn call(
            &mut self,
            msg: Self::CallMessage,
            _context: &Context<Self::Spec>,
            state: &mut impl TxState<S>,
        ) -> Result<(), Error> {
            self.data
                .set(&msg, state)
                .map_err(|e| Error::ModuleError(e.into()))?;
            Ok(())
        }
    }

    pub mod rpc {
        use super::*;
        use crate::my_module::QueryModule;

        #[derive(Debug, Eq, PartialEq, Clone, serde::Serialize, serde::Deserialize)]
        pub struct QueryResponse {
            pub value: Option<String>,
        }

        #[rpc_gen(client, server, namespace = "queryModule")]
        impl<S, D: Data> QueryModule<S, D>
        where
            S: Spec,
        {
            #[rpc_method(name = "queryValue")]
            pub fn query_value(&self, state: &mut ApiStateAccessor<S>) -> RpcResult<QueryResponse> {
                let value = self
                    .data
                    .get(state)
                    .unwrap_infallible()
                    .map(|d| format!("{:?}", d));
                Ok(QueryResponse { value })
            }
        }
    }
}

#[expose_rpc]
#[derive(Default, Genesis, DispatchCall, MessageCodec)]
struct Runtime<T: TestSpec, S: Spec> {
    pub first: my_module::QueryModule<S, T::Data>,
}

#[derive(Default, Clone, PartialEq, Eq)]
struct ActualSpec;

impl TestSpec for ActualSpec {
    type Data = u32;
}

fn main() {}
