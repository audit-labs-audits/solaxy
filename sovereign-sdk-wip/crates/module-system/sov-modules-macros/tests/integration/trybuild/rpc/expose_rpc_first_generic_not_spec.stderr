error[E0277]: the trait bound `T: sov_modules_api::Spec` is not satisfied
   --> tests/integration/trybuild/rpc/expose_rpc_first_generic_not_spec.rs:110:16
    |
110 | struct Runtime<T: TestSpec, S: Spec> {
    |                ^ the trait `sov_modules_api::Spec` is not implemented for `T`
    |
note: required by a bound in `sov_modules_api::Genesis::Spec`
   --> $WORKSPACE/crates/module-system/sov-modules-api/src/module/mod.rs
    |
    |     type Spec: Spec;
    |                ^^^^ required by this bound in `Genesis::Spec`
help: consider further restricting this bound
    |
110 | struct Runtime<T: TestSpec + sov_modules_api::Spec, S: Spec> {
    |                            +++++++++++++++++++++++

error[E0277]: the trait bound `T: sov_modules_api::Spec` is not satisfied
   --> tests/integration/trybuild/rpc/expose_rpc_first_generic_not_spec.rs:109:19
    |
109 | #[derive(Default, Genesis, DispatchCall, MessageCodec)]
    |                   ^^^^^^^ the trait `sov_modules_api::Spec` is not implemented for `T`
    |
    = note: this error originates in the derive macro `Genesis` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider further restricting this bound
    |
110 | struct Runtime<T: TestSpec + sov_modules_api::Spec, S: Spec> {
    |                            +++++++++++++++++++++++

error[E0277]: the trait bound `T: sov_modules_api::Spec` is not satisfied
   --> tests/integration/trybuild/rpc/expose_rpc_first_generic_not_spec.rs:108:1
    |
108 | #[expose_rpc]
    | ^^^^^^^^^^^^^ the trait `sov_modules_api::Spec` is not implemented for `T`
    |
note: required by a bound in `ApiState`
   --> $WORKSPACE/crates/module-system/sov-modules-api/src/rest/mod.rs
    |
    | pub struct ApiState<S: Spec, T = ()> {
    |                        ^^^^ required by this bound in `ApiState`
    = note: this error originates in the attribute macro `expose_rpc` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider further restricting this bound
    |
110 | struct Runtime<T: TestSpec + sov_modules_api::Spec, S: Spec> {
    |                            +++++++++++++++++++++++

error[E0277]: `ActualSpec` doesn't implement `std::fmt::Debug`
   --> tests/integration/trybuild/rpc/expose_rpc_first_generic_not_spec.rs:117:19
    |
117 | impl TestSpec for ActualSpec {
    |                   ^^^^^^^^^^ `ActualSpec` cannot be formatted using `{:?}`
    |
    = help: the trait `std::fmt::Debug` is not implemented for `ActualSpec`
    = note: add `#[derive(Debug)]` to `ActualSpec` or manually `impl std::fmt::Debug for ActualSpec`
note: required by a bound in `TestSpec`
   --> tests/integration/trybuild/rpc/expose_rpc_first_generic_not_spec.rs:10:29
    |
9   | pub trait TestSpec:
    |           -------- required by a bound in this trait
10  |     Default + Send + Sync + std::fmt::Debug + Clone + PartialEq + Eq + 'static
    |                             ^^^^^^^^^^^^^^^ required by this bound in `TestSpec`
help: consider annotating `ActualSpec` with `#[derive(Debug)]`
    |
115 + #[derive(Debug)]
116 | struct ActualSpec;
    |

error[E0277]: the trait bound `T: sov_modules_api::Spec` is not satisfied
   --> tests/integration/trybuild/rpc/expose_rpc_first_generic_not_spec.rs:110:16
    |
110 | struct Runtime<T: TestSpec, S: Spec> {
    |                ^ the trait `sov_modules_api::Spec` is not implemented for `T`
    |
note: required by a bound in `sov_modules_api::DispatchCall::Spec`
   --> $WORKSPACE/crates/module-system/sov-modules-api/src/module/dispatch.rs
    |
    |     type Spec: Spec;
    |                ^^^^ required by this bound in `DispatchCall::Spec`
help: consider further restricting this bound
    |
110 | struct Runtime<T: TestSpec + sov_modules_api::Spec, S: Spec> {
    |                            +++++++++++++++++++++++

error[E0277]: the trait bound `T: sov_modules_api::Spec` is not satisfied
   --> tests/integration/trybuild/rpc/expose_rpc_first_generic_not_spec.rs:109:28
    |
109 | #[derive(Default, Genesis, DispatchCall, MessageCodec)]
    |                            ^^^^^^^^^^^^ the trait `sov_modules_api::Spec` is not implemented for `T`
    |
note: required by a bound in `StateProvider`
   --> $WORKSPACE/crates/module-system/sov-modules-api/src/state/accessors/mod.rs
    |
    | pub trait StateProvider<S: Spec>:
    |                            ^^^^ required by this bound in `StateProvider`
    = note: this error originates in the derive macro `DispatchCall` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider further restricting this bound
    |
110 | struct Runtime<T: TestSpec + sov_modules_api::Spec, S: Spec> {
    |                            +++++++++++++++++++++++

error[E0271]: type mismatch resolving `<QueryModule<S, <T as TestSpec>::Data> as ModuleInfo>::Spec == T`
   --> tests/integration/trybuild/rpc/expose_rpc_first_generic_not_spec.rs:109:19
    |
109 | #[derive(Default, Genesis, DispatchCall, MessageCodec)]
    |                   ^^^^^^^ expected type parameter `T`, found type parameter `S`
110 | struct Runtime<T: TestSpec, S: Spec> {
    |                -            - found type parameter
    |                |
    |                expected type parameter
    |
    = note: expected type parameter `T`
               found type parameter `S`
    = note: a type parameter was expected, but a different one was found; you might be missing a type parameter or trait bound
    = note: for more information, visit https://doc.rust-lang.org/book/ch10-02-traits.html#traits-as-parameters
    = note: required for the cast from `&QueryModule<S, <T as TestSpec>::Data>` to `&dyn sov_modules_api::ModuleInfo<Spec = T>`
    = note: this error originates in the derive macro `Genesis` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `impl ::sov_modules_api::GenesisState<<Self as ::sov_modules_api::Genesis>::Spec>: GenesisState<S>` is not satisfied
   --> tests/integration/trybuild/rpc/expose_rpc_first_generic_not_spec.rs:109:19
    |
109 | #[derive(Default, Genesis, DispatchCall, MessageCodec)]
    |                   ^^^^^^^ the trait `GenesisState<S>` is not implemented for `impl ::sov_modules_api::GenesisState<<Self as ::sov_modules_api::Genesis>::Spec>`
    |
note: required by a bound in `sov_modules_api::Genesis::genesis`
   --> $WORKSPACE/crates/module-system/sov-modules-api/src/module/mod.rs
    |
    |     fn genesis(
    |        ------- required by a bound in this associated function
...
    |         state: &mut impl GenesisState<Self::Spec>,
    |                          ^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Genesis::genesis`
    = note: this error originates in the derive macro `Genesis` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `T: sov_modules_api::Spec` is not satisfied
   --> tests/integration/trybuild/rpc/expose_rpc_first_generic_not_spec.rs:109:28
    |
109 | #[derive(Default, Genesis, DispatchCall, MessageCodec)]
    |                            ^^^^^^^^^^^^ the trait `sov_modules_api::Spec` is not implemented for `T`, which is required by `WorkingSet<T, I>: sov_modules_api::TxState<S>`
    |
    = note: required for `WorkingSet<T, I>` to implement `GasMeter`
    = note: required for `WorkingSet<T, I>` to implement `sov_modules_api::TxState<S>`
note: required by a bound in `sov_modules_api::Module::call`
   --> $WORKSPACE/crates/module-system/sov-modules-api/src/module/mod.rs
    |
    |     fn call(
    |        ---- required by a bound in this associated function
...
    |         _state: &mut impl TxState<Self::Spec>,
    |                           ^^^^^^^^^^^^^^^^^^^ required by this bound in `Module::call`
    = note: this error originates in the derive macro `DispatchCall` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider further restricting this bound
    |
110 | struct Runtime<T: TestSpec + sov_modules_api::Spec, S: Spec> {
    |                            +++++++++++++++++++++++

error[E0271]: type mismatch resolving `<WorkingSet<T, I> as GasMeter>::Spec == S`
   --> tests/integration/trybuild/rpc/expose_rpc_first_generic_not_spec.rs:109:28
    |
109 | #[derive(Default, Genesis, DispatchCall, MessageCodec)]
    |                            ^^^^^^^^^^^^ expected type parameter `S`, found type parameter `T`
110 | struct Runtime<T: TestSpec, S: Spec> {
    |                -            - expected type parameter
    |                |
    |                found type parameter
    |
    = note: expected type parameter `S`
               found type parameter `T`
    = note: a type parameter was expected, but a different one was found; you might be missing a type parameter or trait bound
    = note: for more information, visit https://doc.rust-lang.org/book/ch10-02-traits.html#traits-as-parameters
    = note: required for `WorkingSet<T, I>` to implement `sov_modules_api::TxState<S>`
note: required by a bound in `sov_modules_api::Module::call`
   --> $WORKSPACE/crates/module-system/sov-modules-api/src/module/mod.rs
    |
    |     fn call(
    |        ---- required by a bound in this associated function
...
    |         _state: &mut impl TxState<Self::Spec>,
    |                           ^^^^^^^^^^^^^^^^^^^ required by this bound in `Module::call`
    = note: this error originates in the derive macro `DispatchCall` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
   --> tests/integration/trybuild/rpc/expose_rpc_first_generic_not_spec.rs:109:28
    |
109 | #[derive(Default, Genesis, DispatchCall, MessageCodec)]
    |                            ^^^^^^^^^^^^
    |                            |
    |                            expected `&Context<S>`, found `&Context<T>`
    |                            arguments to this function are incorrect
110 | struct Runtime<T: TestSpec, S: Spec> {
    |                -            - expected type parameter
    |                |
    |                found type parameter
    |
    = note: expected reference `&sov_modules_api::Context<S>`
               found reference `&sov_modules_api::Context<T>`
    = note: a type parameter was expected, but a different one was found; you might be missing a type parameter or trait bound
    = note: for more information, visit https://doc.rust-lang.org/book/ch10-02-traits.html#traits-as-parameters
note: method defined here
   --> $WORKSPACE/crates/module-system/sov-modules-api/src/module/mod.rs
    |
    |     fn call(
    |        ^^^^
    = note: this error originates in the derive macro `DispatchCall` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0271]: type mismatch resolving `<QueryModule<S, <T as TestSpec>::Data> as ModuleInfo>::Spec == T`
   --> tests/integration/trybuild/rpc/expose_rpc_first_generic_not_spec.rs:109:28
    |
109 | #[derive(Default, Genesis, DispatchCall, MessageCodec)]
    |                            ^^^^^^^^^^^^ expected type parameter `T`, found type parameter `S`
110 | struct Runtime<T: TestSpec, S: Spec> {
    |                -            - found type parameter
    |                |
    |                expected type parameter
    |
    = note: expected type parameter `T`
               found type parameter `S`
    = note: a type parameter was expected, but a different one was found; you might be missing a type parameter or trait bound
    = note: for more information, visit https://doc.rust-lang.org/book/ch10-02-traits.html#traits-as-parameters
    = note: required for the cast from `&QueryModule<S, <T as TestSpec>::Data>` to `&dyn sov_modules_api::ModuleInfo<Spec = T>`
    = note: this error originates in the derive macro `DispatchCall` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `<T as TestSpec>::Data: std::default::Default` is not satisfied
   --> tests/integration/trybuild/rpc/expose_rpc_first_generic_not_spec.rs:110:8
    |
110 | struct Runtime<T: TestSpec, S: Spec> {
    |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::default::Default` is not implemented for `<T as TestSpec>::Data`, which is required by `Runtime<T, S>: std::default::Default`
    |
note: required for `Runtime<T, S>` to implement `std::default::Default`
   --> tests/integration/trybuild/rpc/expose_rpc_first_generic_not_spec.rs:109:10
    |
109 | #[derive(Default, Genesis, DispatchCall, MessageCodec)]
    |          ^^^^^^^ unsatisfied trait bound introduced in this `derive` macro
    = note: this error originates in the derive macro `Default` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: the method `clone` exists for struct `ApiState<T>`, but its trait bounds were not satisfied
   --> tests/integration/trybuild/rpc/expose_rpc_first_generic_not_spec.rs:108:1
    |
108 | #[expose_rpc]
    | ^^^^^^^^^^^^^ method cannot be called on `ApiState<T>` due to unsatisfied trait bounds
    |
   ::: $WORKSPACE/crates/module-system/sov-modules-api/src/rest/mod.rs
    |
    | pub struct ApiState<S: Spec, T = ()> {
    | ------------------------------------ doesn't satisfy `ApiState<T>: Clone`
    |
    = note: the following trait bounds were not satisfied:
            `T: sov_modules_api::Spec`
            which is required by `ApiState<T>: Clone`
    = note: this error originates in the attribute macro `expose_rpc` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider restricting the type parameter to satisfy the trait bound
    |
108 | #[expose_rpc] where T: sov_modules_api::Spec
    |               ++++++++++++++++++++++++++++++
